File: 662.py
from typing import Iterator, Tuple
import itertools
import math
import numpy as np
import time


def fibonacci() -> Iterator[int]:
    previous = 0
    current = 1
    while True:
        next = previous + current
        previous = current
        current = next
        yield current


def valid_pythagorean_pairs(max_distance: float) -> Iterator[Tuple[int, int]]:
    ''' Returns each pythagorean pair whose distance is a fibonacci number.

    Dedupes pairs: for example it will yield (3, 4) but won't yield (4, 3).

    max_distance is inclusive.
    '''
    fibonaccis = set(itertools.takewhile(
        lambda _x: _x <= max_distance, fibonacci()))
    # See https://en.wikipedia.org/wiki/Pythagorean_triple#Generating_a_triple
    # for why this algorithm generates each pythagorean triple once.
    #
    # This algorithm is way faster than iterating over a and b directly.
    sqrt_max_distance = math.sqrt(max_distance)
    for m in range(1, int(sqrt_max_distance) + 1):
        m_square = m * m
        for n in range(m):
            if math.gcd(m, n) != 1:
                continue
            if m % 2 == 1 and n % 2 == 1:
                continue
            n_square = n * n
            c = m_square + n_square
            a = m_square - n_square
            b = 2 * m * n
            for k in range(1, max_distance // c + 1):
                if k * c in fibonaccis:
                    yield (k * a, k * b)


MODULO = 1000000007


def num_paths(max_i: int) -> np.ndarray:

    max_distance = int(max_i * math.sqrt(2))
    valid_arcs = list(valid_pythagorean_pairs(max_distance))
    valid_arcs += [tuple(reversed(pair)) for pair in valid_arcs]

    # We split the valid arcs into two groups sorted by x or y, so that we can
    # stop early when we iterate through them.
    valid_arcs_by_y = sorted(
        [a for a in valid_arcs if a[0] < a[1]], key=lambda _p: _p[1])
    valid_arcs_by_x = sorted([a for a in valid_arcs if a[0] >= a[1]])

    t = time.time()

    num_paths_array = []
    for i in range(max_i+1):
        num_paths_array.append([0]*(i+1))
        for j in range(i+1):
            if i == 0 and j == 0:
                num_paths_array[i][j] = 1
                continue
            num_paths_for_i_j = 0

            for arc in valid_arcs_by_y:
                smaller_j = j - arc[1]
                if smaller_j < 0:
                    # The valid arcs are sorted by increasing y, therefore the
                    # next arcs are all going to have y too large as well.
                    #  continue
                    break
                smaller_i = i - arc[0]
                if smaller_i < 0:
                    # Cannot take that arc.
                    continue
                if smaller_j > smaller_i:
                    smaller_i, smaller_j = smaller_j, smaller_i
                num_paths_for_i_j += num_paths_array[smaller_i][smaller_j]

            for arc in valid_arcs_by_x:
                smaller_i = i - arc[0]
                if smaller_i < 0:
                    break
                smaller_j = j - arc[1]
                if smaller_j < 0:
                    continue
                if smaller_j > smaller_i:
                    smaller_i, smaller_j = smaller_j, smaller_i
                num_paths_for_i_j += num_paths_array[smaller_i][smaller_j]

            # Valid_arcs has ~100 elements, so num_paths_for_i_j is roughly
            # 100e9, which fits in an int64. That guarantees that we haven't
            # overflowed so far.
            num_paths_array[i][j] = num_paths_for_i_j % MODULO

    print(f'Time: {time.time() - t}s.')
    #  for i in range(len(num_paths_array)):
    #      for j in range(len(num_paths_array[i])):
    #          print(f'{num_paths_array[i][j]: 11}', end='')
    #      print('')

    return num_paths_array


N = 2000


print(f'Result: {num_paths(N)[N][N]}')

File: 662.cc
#include <cmath>
#include <cstdint>
#include <cstdlib>
#include <iostream>
#include <numeric>
#include <unordered_set>
#include <utility>
#include <vector>

#include "util.h"

struct Arc {
  int x;
  int y;
};

std::vector<Arc> ValidArcs(int width, int height) {
  double max_distance = std::sqrt(width * width + height * height);
  const std::vector<int> fibonacci_numbers_vec =
      Fibonacci(static_cast<int>(max_distance));
  const std::unordered_set<int> fibonacci_numbers(
      fibonacci_numbers_vec.cbegin(), fibonacci_numbers_vec.cend());

  std::vector<Arc> result;
  for (int m = 1; m <= max_distance; ++m) {
    int m2 = m * m;
    for (int n = 0; n < m; ++n) {
      if (m % 2 == 1 && n % 2 == 1) {
        continue;
      }
      if (std::gcd(m, n) != 1) {
        continue;
      }

      int n2 = n * n;
      int c = m2 + n2;
      int a = m2 - n2;
      int b = 2 * m * n;

      for (int k = 1; k <= max_distance / c; ++k) {
        if (fibonacci_numbers.find(k * c) != fibonacci_numbers.cend()) {
          if (k * a <= width && k * b <= height) {
            result.push_back(Arc{.x = k * a, .y = k * b});
          }
          if (k * b <= width && k * a <= height) {
            result.push_back(Arc{.x = k * b, .y = k * a});
          }
        }
      }
    }
  }
  return result;
}

int NumPaths(int width, int height) {
  // NumPaths(i, j) == NumPaths(j, i), so we only compute and store the
  // ones where i <= j. Therefore we set width <= height.
  width = std::min(width, height);
  height = std::max(width, height);

  constexpr int kModulo = 1000000007;
  std::vector<Arc> valid_arcs = ValidArcs(width, height);

  std::vector<std::vector<int>> num_paths;
  num_paths.reserve(width + 1);
  for (int i = 0; i <= width; ++i) {
    num_paths.push_back(std::vector<int>(height + 1));
    for (int j = 0; j <= height; ++j) {
      if (i == 0 && j == 0) {
        num_paths[i][j] = 1;
        continue;
      }

      // Valid_arcs has ~100 elements, so num_paths_for_i_j is roughly 100e9,
      // which fits in an int64. This way, we can do one modulo at the end of
      // the sum rather than for each term in the sum.
      uint64_t num_paths_i_j = 0;
      for (const Arc &arc : valid_arcs) {
        // We iterate over the possible last arcs of the path.
        int smaller_i = i - arc.x;
        int smaller_j = j - arc.y;
        if (smaller_i < 0 || smaller_j < 0) {
          // This arc is too large.
          continue;
        }

        if (smaller_i > smaller_j) {
          // We only stores the ones where i <= j.
          int tmp = smaller_i;
          smaller_i = smaller_j;
          smaller_j = tmp;
        }
        num_paths_i_j += num_paths[smaller_i][smaller_j];
      }
      num_paths[i][j] = static_cast<int>(num_paths_i_j % kModulo);
    }
  }

  return num_paths[width][height];
}

int main() {
  std::cout << "Fibonacci(10000): ";
  for (int i : Fibonacci(10000)) {
    std::cout << i << " ";
  }
  std::cout << std::endl;

  auto valid_arcs = ValidArcs(100, 100);
  std::cout << "ValidArcs(100, 100): ";
  for (const Arc &arc : valid_arcs) {
    std::cout << "(" << arc.x << ", " << arc.y << ") ";
  }
  std::cout << std::endl;

  std::cout << "NumPaths(3, 4) = " << NumPaths(3, 4) << std::endl;
  std::cout << "NumPaths(10, 10) = " << NumPaths(10, 10) << std::endl;
  std::cout << "NumPaths(10000, 10000) = " << NumPaths(10000, 10000)
            << std::endl;
  return EXIT_SUCCESS;
}
